### PR TITLE
[Glimmer2] Rerender on run loop if data has changed

### PR DESCRIPTION
**Pull request by nickiaconis**
Marking an object as dirty kicks off a run loop if one has not already been started. At the start of every run loop, a task is scheduled on the render queue to rerender the main view if any changes have occurred.

Tested by disabling the automatic invocation of manual rerendering in the glimmer tests.

Will require ember-dev to be updated since this invalidates the assumption that a run loop should never be automatically started inside a test.

### CODE REVIEW

File: packages/ember-glimmer/lib/ember-metal-views/index.js [hunk 1]
```diff
@@ -13,6 +15,17 @@ class DynamicScope {
   }
 }
 
+function setupRerenderLoop(renderer, view) {
+  // let LAST_TAG_VALUE = CURRENT_TAG.value();
+  run.backburner.on('begin', () => {
+    run.backburner.schedule('render', () => {
```
Comment (krisselden) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: we don't need to recreate this anon function every time
it also needs to go away when the root view is gone, maybe the renderer should track root views.

File: packages/ember-metal/lib/tags.js [hunk 2]
```diff
@@ -1,5 +1,7 @@
 import { meta as metaFor } from './meta';
 import require, { has } from 'require';
+import run from './run_loop';
+import { K } from './core';
```
Comment (krisselden) [inline packages/ember-metal/lib/tags.js]: just define `function K() {}`

File: packages/ember-glimmer/lib/ember-metal-views/index.js (hunk 1 above)
Comment (nickiaconis) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: So the Renderer's appendOutletView and appendTo methods would add a view to the list, and its remove method would remove that view? Then, when a view is removed, if the list is empty the begin handler is unregistered from the run loop?

File: packages/ember-glimmer/lib/ember-metal-views/index.js [hunk 3]
```diff
@@ -61,6 +99,7 @@ class Renderer {
   }
 
   remove(view) {
+    deregisterViewAutorender(view);
```
Comment (krisselden) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: does this get called while looping over RENDERED_VIEWS or out of band?

File: packages/ember-glimmer/lib/ember-metal-views/index.js [hunk 4]
```diff
@@ -13,6 +15,38 @@ class DynamicScope {
...
 }
 
+const RENDERED_VIEWS = [];
+// let LAST_TAG_VALUE;
+
+function maybeRerender() {
+  // if (CURRENT_TAG.validate(LAST_TAG_VALUE)) { return; }
+  for (let i = 0; i < RENDERED_VIEWS.length; i += 2) {
+    let view = RENDERED_VIEWS[i];
+    let renderer = RENDERED_VIEWS[i + 1];
```
Comment (krisselden) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: renderer is injected onto the root view

File: packages/ember-glimmer/lib/ember-metal-views/index.js [hunk 5]
```diff
@@ -37,6 +71,8 @@ class Renderer {
     let result = view.template.asEntryPoint().render(self, env, { appendTo: target, dynamicScope });
     env.commit();
 
+    registerViewAutorender(view, this);
```
Comment (krisselden) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: just make it registerView

File: packages/ember-glimmer/lib/ember-metal-views/index.js [hunk 6]
```diff
@@ -13,6 +15,38 @@ class DynamicScope {
...
+  // if (CURRENT_TAG.validate(LAST_TAG_VALUE)) { return; }
+  for (let i = 0; i < RENDERED_VIEWS.length; i += 2) {
+    let view = RENDERED_VIEWS[i];
+    let renderer = RENDERED_VIEWS[i + 1];
+    if (!view.isDestroyed) { renderer.rerender(view); }
+  }
+  // LAST_TAG_VALUE = CURRENT_TAG.value();
+}
+
+function scheduleRerender() {
```
Comment (krisselden) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: rerender and renderer  visually looks similar, maybe scheduleMaybeUpdate and maybeUpdate?

File: packages/ember-metal/lib/tags.js [hunk 7]
```diff
@@ -19,11 +20,19 @@ export function tagFor(object, _meta) {
   }
 }
 
+function K() {}
+function ensureRunloop() {
+  if (!run.backburner.currentInstance) {
```
Comment (krisselden) [inline packages/ember-metal/lib/tags.js]: `const { backburner } = run;` up top?

File: packages/ember-metal/lib/tags.js [hunk 8]
```diff
@@ -19,11 +20,19 @@ export function tagFor(object, _meta) {
...
+    run.backburner.schedule('actions', K);
+  }
+}
+
 if (hasGlimmer) {
   ({ DirtyableTag, CONSTANT_TAG, CURRENT_TAG } = require('glimmer-reference'));
   makeTag = function() { return new DirtyableTag(); };
 
   markObjectAsDirty = function(meta) {
+    ensureRunloop();
```
Comment (krisselden) [inline packages/ember-metal/lib/tags.js]: since the current ember-glimmer test harness always starts a run loop with runTask() we could do this in a followup PR, just focus on getting the other part in place and validating the tag.

File: packages/ember-glimmer/lib/ember-metal-views/index.js [hunk 9]
```diff
@@ -61,6 +98,7 @@ class Renderer {
   }
 
   remove(view) {
```
Comment (krisselden) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: is this only called on a top level view?

File: packages/ember-glimmer/lib/ember-metal-views/index.js [hunk 10]
```diff
@@ -13,6 +15,37 @@ class DynamicScope {
   }
 }
 
+const RENDERED_VIEWS = [];
```
Comment (krisselden) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: should name this clearly, like RENDERED_ROOTS; want to ensure people realize this are roots and not any view.

File: packages/ember-glimmer/tests/utils/test-case.js [hunk 11]
```diff
@@ -31,10 +31,10 @@ export class RenderingTest extends AbstractRenderingTest {
     this.renderer._root = this.component;
   }
 
-  runTask(callback) {
-    super.runTask(() => {
-      callback();
-      this.component.rerender();
-    });
-  }
+  // runTask(callback) {
```
Comment (krisselden) [inline packages/ember-glimmer/tests/utils/test-case.js]: just remove this.

File: packages/ember-metal/lib/tags.js [hunk 12]
```diff
@@ -1,6 +1,8 @@
 import { meta as metaFor } from './meta';
 import require, { has } from 'require';
+// import run from './run_loop';
```
Comment (krisselden) [inline packages/ember-metal/lib/tags.js]: better to pull this up into its own commit based on this PR

File: packages/ember-glimmer/lib/ember-metal-views/index.js [hunk 13]
```diff
@@ -13,12 +15,41 @@ class DynamicScope {
...
 class Renderer {
   constructor({ dom, env, destinedForDOM = false }) {
     this._root = null;
     this._dom = dom;
     this._env = env;
     this._destinedForDOM = destinedForDOM;
+    this._renderedRoots = [];
+    this._scheduler = function scheduleMaybeUpdate() {
+      run.backburner.schedule('render', null, maybeUpdate, CURRENT_TAG.value(), this._renderedRoots);
+    }.bind(this);
```
Comment (krisselden) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: Just use () => {} instead of bind

Comment (krisselden) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: Could also just make this the second arg too and put maybeUpdate on the renderer and not pass the args but just use this

File: packages/ember-glimmer/lib/ember-metal-views/index.js [hunk 14]
```diff
@@ -13,12 +15,41 @@ class DynamicScope {
...
+  }
+}
+
 class Renderer {
   constructor({ dom, env, destinedForDOM = false }) {
     this._root = null;
     this._dom = dom;
     this._env = env;
     this._destinedForDOM = destinedForDOM;
+    this._renderedRoots = [];
```
Comment (krisselden) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: Make this an object with register and unregister methods

File: packages/ember-glimmer/lib/ember-metal-views/index.js (hunk 13 above)
Comment (stefanpenner) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: could we make the caller instead support the following (that way we don't need to retain the context, or rebind later?

```js
invoke(schedulable)

function invoke(x) {
  // later invoke x on the method in schedulable
  x.scheduleMaybeUpdate()
}
```

Comment (nickiaconis) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: @stefanpenner I don't understand how this helps with the fact that the callback's context is lost once it is passed to `backburner.on`. Could you explain?

Comment (stefanpenner) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: sorry, i was mostly thinking outloud, I think it would be great if someday we merely have some nice protocol to avoid that.

**Comment by nickiaconis** [comment]
Kicking Travis...

File: packages/ember-glimmer/lib/ember-metal-views/index.js (hunk 13 above)
Comment (nickiaconis) [inline packages/ember-glimmer/lib/ember-metal-views/index.js]: Something like how you can pass a context to schedule? That would be nice.

**Comment by chadhietala** [comment]
LGTM. Are there any tests that we can either flip to `@test` or ones we could write to validate? Maybe `this.runTask` stuff can now be cleaned up?

**Comment by nickiaconis** [comment]
I think `this.runTask` needs to stay for now. The changes here rerender on
the run loop, but we'll need to ensure a run loop is running whenever
something is set, which is not included here, before we can strip the
`this.runTask` stuff.
